File: Algorithms/p023_Merge_k_Sorted_Lists/p023_Merge_k_Sorted_Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        ListNode head(0);
        ListNode* tail = &head;
        auto cmp = [&](ListNode* a, ListNode* b){ return a->val>b->val; };
        priority_queue<ListNode*,vector<ListNode*>,decltype(cmp)> heap(cmp);
        for(auto e : lists){
            if(e) heap.push(e);
        }
        while(!heap.empty()){
            tail->next = heap.top();
            heap.pop();
            tail = tail->next;
            if(tail->next) heap.push(tail->next);
        }
        return head.next;
    }
};

File: Algorithms/p023_Merge_k_Sorted_Lists/p023_Merge_k_Sorted_Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        head = ListNode(None)
        tail = head
        heap = []
        for e in lists:
            if e:
                heapq.heappush(heap,(e.val,e))
        while heap:
            tail.next = heap[0][1]
            heapq.heappop(heap)
            tail = tail.next
            if tail.next:
                heapq.heappush(heap,(tail.next.val,tail.next))
        return head.next
